riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1822 @@
+    // stores the lower 32 bits in register Z and the upper 32 bits in register W.
+    instr mul XL, YL, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> (tmp4_h, tmp4_l, tmp3_h, tmp3_l) = arith_mul.mul(tmp1_h, tmp1_l, 0, 0, tmp2_h, tmp2_l)
+        link ~> regs.mstore(0, ZL, STEP + 2, tmp3_h, tmp3_l)
+        link ~> regs.mstore(0, WL, STEP + 3, tmp4_h, tmp4_l);
+"#
+}
+
+fn memory(with_bootloader: bool) -> String {
+    let memory_machine = if with_bootloader {
+        todo!()
+    } else {
+        r#"
+    std::machines::small_field::memory::Memory memory(bit12, byte2);
+"#
+    };
+
+    memory_machine.to_string()
+        + r#"
+
+    // Increased by 4 in each step, because we do up to 4 register memory accesses per step
+    col fixed STEP(i) { 4 * i };
+
+    // ============== memory instructions ==============
+
+    /// Loads one word from an address V = val(X) + Y and rounds it down to the next multiple of 4.
+    /// Writes the loaded word and the remainder of the division by 4 to registers Z and W,
+    /// respectively.
+    instr mload XL, YH, YL, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+
+        link ~> (tmp2_h, tmp2_l) = add_sub.add(tmp1_h, tmp1_l, YH, YL)
+
+        link ~> (tmp3_h, tmp3_l) = memory.mload(tmp2_h, X_b2 * 0x100 + X_b1 * 4, STEP + 1)
+        link ~> regs.mstore(0, ZL, STEP + 2, tmp3_h, tmp3_l)
+        link ~> regs.mstore(0, WL, STEP + 3, 0, tmp4_l)
+        link => bit2.check(tmp4_l)
+        link => bit6.check(X_b1)
+        link => byte.check(X_b2)
+    {
+        tmp2_l = X_b2 * 0x100 + X_b1 * 4 + tmp4_l
+    }
+
+    // Stores val(W) at address (V = val(X) - val(Y) + Z) % 2**32.
+    // V can be between 0 and 2**33.
+    // V should be a multiple of 4, but this instruction does not enforce it.
+    instr mstore XL, YL, ZH, ZL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> (tmp2_h, tmp2_l) = regs.mload(0, YL, STEP + 1)
+        link ~> (tmp3_h, tmp3_l) = regs.mload(0, WL, STEP + 2)
+
+        link ~> (tmp4_h, tmp4_l) = add_sub.sub(tmp1_h, tmp1_l, tmp2_h, tmp2_l)
+        link ~> (tmp5_h, tmp5_l) = add_sub.add(tmp4_h, tmp4_l, ZH, ZL)
+
+        link ~> memory.mstore(tmp5_h, tmp5_l, STEP + 3, tmp3_h, tmp3_l);
+
+"#
+}
+
+fn only_if_no_write_to_zero(reg: Register, statement: String) -> Vec<String> {
+    only_if_no_write_to_zero_vec(reg, vec![statement])
+}
+
+fn only_if_no_write_to_zero_vec(reg: Register, statements: Vec<String>) -> Vec<String> {
+    if reg.is_zero() {
+        vec![]
+    } else {
+        statements
+    }
+}
+
+/// Push register into the stack
+
+pub fn push_register(name: &str) -> Vec<String> {
+    assert!(name.starts_with('x'), "Only x registers are supported");
+    let reg = Register::from(name);
+    vec![
+        // x2 + x0 - 4 => x2
+        format!("add_wrap 2, 0, 0xffff, 0xfffc, 2;",),
```

Comment:
Would it be useful to have a utility function to turn -4 into (0xffff, 0xfffc)?

Reply:
Based on `i32_high(-4), i32_low(-4)` ?

Reply:
done